### PR TITLE
when kubelet removes pod dir, it should skip those that still have unmounted volumes

### LINKED ISSUE #12155: kubernetes deletes info from all NFS shares when one of the mounts becames stalled (google container engine)
When mounting several shares over NFS to container and when one of the directories is deleted from NFS server, something in the node is deleting content from other shares. I suspect that something goes wrong with "cleanupOrphanedVolumes"? not sure tough.

Use case: php-app is running in container, uses php-sessions for remembering login. php sessions are stored in file and shared through nfs-share between containers. also app code is shared and several other directories to store different data (data dirs are mapped under code, code is in separate share so generated document files don't get to git tree), all is managed by separate integration server (in container), that uses same shares to prepare the code.

This happened first time in GCE with Cluster API version 0.17.something, now it acts same on 1.0.1 and 0.18.2. Problem is repeatable:
First of all you need to install package "nfs-common" to **all cluster nodes**, otherwise you can't map NFS shares at all:

    apt-get update && apt-get install nfs-common

now you need nfs server. I created small instance for that with /etc/export stating:

    /export 	10.240.0.0/16(rw,sync,crossmnt,no_subtree_check)

also I made directories under /export/test/:
 
    #ls -la ubuntutest ubuntutest2 ubuntutest3
    ubuntutest:
    total 12
    drwxrwxrwx  2 root root 4096 Aug  3 15:07 .
    drwxr-xr-x 12 root root 4096 Aug  3 15:06 ..
    -rw-r--r--  1 root root 1580 Aug  3 15:06 testifail
    ubuntutest2:
    total 12
    drwxrwxrwx  2 root root 4096 Aug  3 15:06 .
    drwxr-xr-x 12 root root 4096 Aug  3 15:06 ..
    -rw-r--r--  1 root root 1580 Aug  3 15:06 testifail2
    ubuntutest3:
    total 12
    drwxr-xr-x  2 root root 4096 Aug  3 15:07 .
    drwxr-xr-x 12 root root 4096 Aug  3 15:06 ..
    -rw-r--r--  1 root root 1580 Aug  3 15:07 testifail3

next you need ubuntutest.yaml to start it up (replace nfs server ip with your own):

    apiVersion: v1
    kind: ReplicationController
    metadata:
      labels:
        name: zubuntu
      name: zubuntu
    spec:
      replicas: 1
      selector:
        name: zubuntu-rc
      template:
        metadata:
          labels:
            name: zubuntu-rc
        spec:
          restartPolicy: Always
          containers:
            - image: tutum/apache-php
              name: ubuntu-cont
              ports:
                - containerPort: 80
                  name: web-port
              volumeMounts:
                - name: test-nfs-1
                  mountPath: /app
                - name: test-nfs-2
                  mountPath: /app2
                - name: test-nfs-3
                  mountPath: /app3
          volumes:
            - name: test-nfs-1
              nfs:
                server: 10.240.175.237
                path: /export/test/ubuntutest
            - name: test-nfs-2
              nfs:
                server: 10.240.175.237
                path: /export/test/ubuntutest2
            - name: test-nfs-3
              nfs:
                server: 10.240.175.237
                path: /export/test/ubuntutest3
    
now start up the pod:

    kubectl create -f ubuntutest.yaml 
    kubectl get pods

verify, that the directories are mapped:

    kubectl exec -ti zubuntu-8y55c bash
	# ls -la /app*
	/app:
	total 12
	drwxrwxrwx  2 root root 4096 Aug  3 18:07 .
	drwxr-xr-x 69 root root 4096 Aug  3 18:09 ..
	-rw-r--r--  1 root root 1580 Aug  3 18:06 testifail

	/app2:
	total 12
	drwxrwxrwx  2 root root 4096 Aug  3 18:06 .
	drwxr-xr-x 69 root root 4096 Aug  3 18:09 ..
	-rw-r--r--  1 root root 1580 Aug  3 18:06 testifail2

	/app3:
	total 12
	drwxr-xr-x  2 root root 4096 Aug  3 18:07 .
	drwxr-xr-x 69 root root 4096 Aug  3 18:09 ..
	-rw-r--r--  1 root root 1580 Aug  3 18:07 testifail3

now the funny part:
remove directory from nfs server

     # rm -rf ubuntutest

the zubuntu pod shows:

	# ls -la /app*
	/app:
	ls: reading directory /app: Stale file handle
	total 0

	/app2:
	total 12
	drwxrwxrwx  2 root root 4096 Aug  3 18:06 .
	drwxr-xr-x 69 root root 4096 Aug  3 18:09 ..
	-rw-r--r--  1 root root 1580 Aug  3 18:06 testifail2

	/app3:
	total 12
	drwxr-xr-x  2 root root 4096 Aug  3 18:07 .
	drwxr-xr-x 69 root root 4096 Aug  3 18:09 ..
	-rw-r--r--  1 root root 1580 Aug  3 18:07 testifail3

nothing happens on nfs server yet, but just launch the stop command:

    kubectl stop -f ubuntutest.yaml 

and now nfs server shows:

	# ls -la ubuntutest ubuntutest2 ubuntutest3
	ls: cannot access ubuntutest: No such file or directory
	ubuntutest2:
	total 8
	drwxrwxrwx  2 root root 4096 Aug  3 15:18 .
	drwxr-xr-x 11 root root 4096 Aug  3 15:17 ..
	ubuntutest3:
	total 12
	drwxr-xr-x  2 root root 4096 Aug  3 15:07 .
	drwxr-xr-x 11 root root 4096 Aug  3 15:17 ..
	-rw-r--r--  1 root root 1580 Aug  3 15:07 testifail3

you see, that one file was left, but not for long. this is because root in client does not have root access in nfs-file system, but just after giving rights, the file is gone:

	# chmod 777 ubuntutest3

	# ls -la ubuntutest ubuntutest2 ubuntutest3
	ls: cannot access ubuntutest: No such file or directory
	ubuntutest2:
	total 8
	drwxrwxrwx  2 root root 4096 Aug  3 15:18 .
	drwxr-xr-x 11 root root 4096 Aug  3 15:17 ..
	ubuntutest3:
	total 8
	drwxrwxrwx  2 root root 4096 Aug  3 15:20 .
	drwxr-xr-x 11 root root 4096 Aug  3 15:17 ..

now you can test "black hole" by creating new files. they dissaperar from the server in few moments. the dissapearing continues until cluster node is restarted.

### PR DESCRIPTION
**Pull request by rootfs**
this is to fix #12155 
@saad-ali @yujuhong

### CODE REVIEW

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1392,7 +1392,7 @@ func (kl *Kubelet) cleanupOrphanedPodDirs(pods []*api.Pod, runningPods []*kubeco
 			continue
 		}
 		glog.V(3).Infof("Orphaned pod %q found, removing", uid)
-		if err := os.RemoveAll(kl.getPodDir(uid)); err != nil {
+		if err := util.RemoveAllSkipMountPoints(kl.getPodDir(uid)); err != nil {
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Hmm... kubelet already checks whether [there exists any not-yet-cleaned-up volume in the pod directory](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1423), and skip the pod directory removal if that's true. Correct me if I am wrong, but I think your PR handles pretty much the same thing?

Comment (rootfs) [inline pkg/kubelet/kubelet.go]: I see, you beat me :) let me read it again.

**Comment by yujuhong** [comment]
@rootfs, I am reopening this to fix #13862.

This is an issue that had always been there, and my fix didn't help :)
Since your fix is going to subsume mine, feel free to delete the old logic.

**Comment by rootfs** [comment]
@yujuhong A bit hacky :) 
so the idea is to get the `ReadDir` resilient to errors so we can get the directory names that are safe to remove later.

**Comment by yujuhong** [comment]
> so the idea is to get the ReadDir resilient to errors so we can get the directory names that are safe to remove later.

How about we skip the directory with an error [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumes.go#L150), continue to read all other directories, and return them with an error? The caller can then decide what to do.

**Comment by rootfs** [comment]
@yujuhong along that line, how about `getPodVolumes` returns both errors and directory list (the new `ReadDirNoExit` needs to change too)? If the errors are something we can accept (like stale handle, transport down, busy, etc, we have to work out the list), then we allow cleanup to continue?

**Comment by yujuhong** [comment]
> @yujuhong along that line, how about getPodVolumes returns both errors and directory list (the new ReadDirNoExit needs to change too)? If the errors are something we can accept (like stale handle, transport down, busy, etc, we have to work out the list), then we allow cleanup to continue?

That sounds good to me. Just to make sure we're on the same page, this should eliminate the need to add `RemoveAllSkipMountPoints`, right?

**Comment by rootfs** [comment]
@yujuhong very likely.

**Comment by rootfs** [comment]
@yujuhong please take a look, thanks.

File: pkg/kubelet/volumes.go [hunk 2]
```diff
@@ -145,7 +145,7 @@ func (kl *Kubelet) getPodVolumes(podUID types.UID) ([]*volumeTuple, error) {
 	for _, volumeKindDir := range volumeKindDirs {
 		volumeKind := volumeKindDir.Name()
 		volumeKindPath := path.Join(podVolDir, volumeKind)
-		volumeNameDirs, err := ioutil.ReadDir(volumeKindPath)
+		volumeNameDirs, err := util.ReadDirNoExit(volumeKindPath)
```
Comment (yujuhong) [inline pkg/kubelet/volumes.go]: nit: it's worth adding a comment here, explaining that it may skip some directories.

File: pkg/util/util.go [hunk 3]
```diff
@@ -519,3 +519,35 @@ func FileExists(filename string) (bool, error) {
...
+	names, err := f.Readdirnames(-1)
+	list := make([]os.FileInfo, 0, len(names))
+	for _, filename := range names {
+		fip, lerr := os.Lstat(dirname + "/" + filename)
+		if os.IsNotExist(lerr) {
+			// File disappeared between readdir + stat.
+			// Just treat it as if it didn't exist.
+			continue
+		}
+		// if lstat fails, fip is nil
```
Comment (yujuhong) [inline pkg/util/util.go]: Can we still return the error somehow? I am worried about silently failing without surfacing any error in kubelet's log.

**Comment by rootfs** [comment]
@yujuhong done

**Comment by yujuhong** [comment]
Thanks! LGTM. Shippable failure seems to be a flake, so I kicked it off again.

**Pull request merged**